### ISSUE TITLE
Timezone for UTC times should be UTC.

### ISSUE THREAD

**Issue by judgej** [calcinai/xero-php#79]
Fetching a transaction I see this in the transaction:

                        [UpdatedDateUTC] => DateTime Object
                                (
                                    [date] => 2015-08-17 12:34:28.260000
                                    [timezone_type] => 3
                                    [timezone] => Europe/London
                                )

I'm guessing the timezone comes from the account timezone, but this timestamp from Xero is special - it is a UTC time and should not be presented in the timezone of the account.

So far as I can see the "date" element is correct if interpreted as UTC, but wrong if interpreted as a "Europe/London" timezone time.

**Comment by judgej** [comment]
The timezone might even be being picked up locally. I've not investigated where it is coming from yet, just noting that what is presented by the API is wrong - the time is correct, but only as a UTC time.

**Comment by judgej** [comment]
By setting the local default timezone, I get this:

    date_default_timezone_set('UTC');

    [UpdatedDateUTC] => DateTime Object
                (
                    [date] => 2015-08-17 12:34:28.260000
                    [timezone_type] => 3
                    [timezone] => UTC
                )

Notice the timezone in the timestamp returned by the API is now UTC, but the time string is not. By setting my local time, this means the time the API library is giving me, is *completely different* - it has jumped backwards by an hour.

It is important to keep the timezone of the remote Xero account when interpreting times/date values that don't have one explicitly in the API data. In this case the API field is *specifically* giving us a UTC time, so that meeds to be handled by the API library.

Dates are going to be a whole other walk in the park, since PHP does not have a date type. I'll come back to those later.

**Comment by judgej** [comment]
Here is the raw API data for the last update time:

    <UpdatedDateUTC>2015-08-17T12:34:28.23</UpdatedDateUTC>

Only the name gives us any indication of the timezone. *However* that may be a Xero bug - other timestamps in the API *do* give us the timezone. For example the current time on the organisation:

    <DateTimeUTC>2015-08-17T17:19:07.5133435Z</DateTimeUTC>

Note the "Z" on the end - that's UTC, that is. See http://stackoverflow.com/questions/8405087/what-is-this-date-format-2011-08-12t201746-384z

Raw data for dates on the same API fetch of the payment:

        <Date>2015-08-10T00:00:00</Date>
        <DueDate>2015-08-27T00:00:00</DueDate>

Again, no indication of a timezone. IMO those dates should not have a time component at all, but they do. So the timezone must be interpreted as that timezone the account is set as. I can see this in my organisation:

    <Timezone>GMTSTANDARDTIME</Timezone>

This equates to "Dublin, Edinburgh, London" in the Xero financial settings screen. That's a bloody mess TBH - we are on DST right now in the summer - NOT GMT. This feels like a horrible mess in Xero's timezone handling. Never mind - if the settings say the account is running on GMT, then this is how the times (without a Z and without UCT in the name) and the dates without times, should be interpreted. If I want to display those times as BST, then I can do that conversion my end, *knowing* that I am getting UST (aka GMT) from Xero through the API.

Hope that all makes sense! I'm going to raise a few issues with Xero now...

**Comment by calcinai** [comment]
What a mess!  Yeah, I didn't actually pay much attention to the timezones that were returned from Xero.  I'm wondering if we just make all ```DateTime``` constructors assume UTC?  

eg: ```new DateTime(x, new DateTimeZone('UTC'));```

**Comment by judgej** [comment]
I've asked Xero for clarification and I'll do a bit more research into the docs now. What needs to be interpreted are:

* Times in a field with UTC in the name, but no trailing "Z".
* Times in a field with UTC in the name, and with a trailing "Z".
* Dates that should not involve a timezone, but are sent as a timestamp with a zero time (no trailing "Z" that I have found so far).

**Comment by judgej** [comment]
This page implicitly confirms that times *AND* dates are set to local time in Xero, local to the account:

http://developer.xero.com/documentation/advanced-docs/dates-and-timestamps-explained/

This is the key part:

> e.g. If an invoice is created for a New Zealand based organisation (with a timezone setting of UTC+12:00) at 9pm 7 Feb 2013 UTC and the date element is not specified then the Date will be set as 8 Feb 2013.

It is still a bit ambiguous to me through. In New Zealand the date will be *set* to 8 Feb, because that is what date it is in 8 Feb. However, is that talking about how that date looks to the New Zealand user through the admin screens? Since the *time* is always set to zero, I can't see how it can be anything other than a local date. It could not be stored as a UTC date, as that would involve the time needing to be set to 12:00:00, which it can't, so the date that we see though the API (8 Feb 2013 00:00:00) *must* be interpreted as being in the timezone of that account (+12). If desired, we can convert that to UTC, which would take it back 12 hours to 7 Feb 2013 12:00:00.

So my conclusion is that dates in Xero (not timestamps - just *dates*) if converted to a PHP `DateTime` data type, MUST be converted using the account timezone. In many cases, if you are just using the `DateTime` to carry a date for formatting, and ignoring the time, then it will make no difference. But if you start doing any arithmetic on the date, then things can get hairy pretty quickly if the correct timezones are not set.

**Comment by judgej** [comment]
Gawd help any Xero user who changes their timezone after running the system for a while. Hopefully Xero stores all dates and times as UTC and just does a last-minute conversion when pulling data out.

**Comment by judgej** [comment]
These are the timezones supported by Xero:

https://github.com/XeroAPI/XeroAPI-Schemas/blob/master/v2.00/Timezone.xsd

It it some kind of Microsoft .net thing, and seems to completely disregard DST (e.g. to say that a time is UTC+2, you must specify a date to qualify it, since with DST it may be +2 sometimes, and +3 other times). Urgh, horrible.

**Comment by judgej** [comment]
Sorry for spamming your inbox!

What I think this means, is that when configuring a php-xero application, it needs to be given the timezone to interpret the dates and times as. That should be used when generating the `DateTime` objects. However, DST needs to be turned off somehow, because it does not look like Xero does DST in its dates and times. [PHP does have a bunch of timezones](http://php.net/manual/en/timezones.others.php) that do not have DST, and should all map directly on to the Xero timezones.

**Comment by judgej** [comment]
Got a reply from Xero confirming that the "Z" is consistently put on the end of all UTC timestamps. I have replied pointing out (again) that it is not - it is only added to half of them, even those with "UTC" in the field name. Taking him at his word, the <DateCreatedUTC> fields are not actually UTC. I'll keep pushing for this to be fixed.

Now, on dates. Running some tests against due dates for invoices over a number of timezones (TUC, TUC-3, UTC+3) it seems that the date returned by the API, in "YYYY-MM-DDT00:00:00.0" format, will be the data *as entered by the user*. It does not matter what the user's organisation timezone was at the time they entered it, or what timezone the organisation is set to now (yikes!) the date returned will not change - it will be as typed by that user, so it will represent the calendar day in the timezone of that user (which should be the timezone set for that organisation at the time, and hopefully still is the same timezone).

**Comment by calcinai** [comment]
Thanks heaps for following that ip.  I've got a couple of ideas to differentiate between the different "types" of dates.  Will sort it out soon.

**Comment by judgej** [comment]
Xero are bloody unbelievable! I copied and pasted those dates above with the "Z" suffix from the API inspector - that was raw data - two days ago. Yesterday a dev replied stating that the Z suffix was used consistently across all the API end points. I reply to point out that it it not consistent, but a bit random. Then today he denies he ever said that, and that he said "Z" suffixes are never used, ever, and never have been. And lo-behold I check the API inspector today and they are gone.

If that had a friggin bug, they should just say, "oh yeah, thanks, fixed it". But making me out to be a lier with a trail of evidence to the contrary, reeks of a very immature development team. It's pathetic TBH.

Anyway - on the bright side, they *now* confirm the Z suffix is not (or no longer!) there, and that the "UTC" in the field name is the definitive indicator that the field contains a UTC timestamp. Solved!
